File: frontend/src/api/healthMonitor.ts
/*
 * PacketArch — OT Traffic Simulation Platform
 * Copyright (c) 2026 Rocky Smith <rocky.d.smith@proton.me>
 * Licensed under GPL-3.0. See LICENSE at the repo root.
 */
/**
 * Health Monitor API Client
 * Endpoints for agent health monitoring, events, and configuration
 */

import apiClient from './client';

export interface HealthEvent {
  id: string;
  timestamp: string;
  event_type: string;
  severity: 'info' | 'warning' | 'critical';
  agent_id: string;
  agent_name: string;
  scenario_id: string | null;
  message: string;
  details: Record<string, unknown>;
  acknowledged: boolean;
}

export interface HealthEventsResponse {
  events: HealthEvent[];
  total: number;
  unacknowledged_count: number;
  counts_by_severity: { info: number; warning: number; critical: number };
}

export interface AgentHealthDetail {
  status: 'healthy' | 'warning' | 'critical' | 'offline';
  heartbeat_ok: boolean;
  resource_ok: boolean;
  stalled_scenarios: string[];
  recent_recoveries: number;
}

export interface HealthStatusResponse {
  agents: Record<string, AgentHealthDetail>;
  summary: { healthy: number; warning: number; critical: number; offline: number };
  auto_recovery_enabled: boolean;
  auto_redeploy_on_reconnect: boolean;
  monitoring_active: boolean;
}

export interface HealthMonitorConfig {
  check_interval_seconds: number;
  heartbeat_timeout_seconds: number;
  stall_detection_seconds: number;
  stall_grace_period_seconds: number;
  resource_warning_threshold: number;
  resource_critical_threshold: number;
  resource_sustained_seconds: number;
  max_recovery_attempts_per_hour: number;
  recovery_cooldown_seconds: number;
  auto_recovery_enabled: boolean;
  auto_redeploy_on_reconnect: boolean;
}

export interface DashboardHealthData {
  agent_statuses: Record<string, 'healthy' | 'warning' | 'critical' | 'offline'>;
  recent_events: HealthEvent[];
  unacknowledged_count: number;
}

const PREFIX = '/api/v1/health-monitor';

export const healthMonitorApi = {
  async getEvents(limit = 50, offset = 0, severity?: string): Promise<HealthEventsResponse> {
    const params = new URLSearchParams({ limit: String(limit), offset: String(offset) });
    if (severity) params.append('severity', severity);
    const res = await apiClient.get<HealthEventsResponse>(`${PREFIX}/events?${params}`);
    return res.data;
  },

  async getStatus(): Promise<HealthStatusResponse> {
    const res = await apiClient.get<HealthStatusResponse>(`${PREFIX}/status`);
    return res.data;
  },

  async acknowledgeEvent(eventId: string): Promise<void> {
    await apiClient.post(`${PREFIX}/events/${eventId}/acknowledge`);
  },

  async getConfig(): Promise<HealthMonitorConfig> {
    const res = await apiClient.get<HealthMonitorConfig>(`${PREFIX}/config`);
    return res.data;
  },

  async updateConfig(config: Partial<HealthMonitorConfig>): Promise<HealthMonitorConfig> {
    const res = await apiClient.put<HealthMonitorConfig>(`${PREFIX}/config`, config);
    return res.data;
  },

  async clearEvents(): Promise<void> {
    await apiClient.delete(`${PREFIX}/events`);
  },
};

File: backend/app/api/routes/health_monitor.py
# PacketArch — OT Traffic Simulation Platform
# Copyright (c) 2026 Rocky Smith <rocky.d.smith@proton.me>
# Licensed under GPL-3.0. See LICENSE at the repo root.
"""Health monitoring API endpoints."""

import logging
from typing import Any
from uuid import UUID

from fastapi import APIRouter

from app.core.exceptions import NotFoundError
from app.services.health_monitor import health_monitor

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/health-monitor", tags=["health-monitor"])


@router.get("/events")
async def get_health_events(
    limit: int = 50,
    offset: int = 0,
    severity: str | None = None,
    agent_id: str | None = None,
) -> dict[str, Any]:
    """Get health events, most recent first.

    Args:
        limit: Max events to return (default 50)
        offset: Skip this many events
        severity: Filter by severity (info, warning, critical)
        agent_id: Filter by agent UUID
    """
    events = health_monitor.get_events(
        limit=limit,
        offset=offset,
        severity_filter=severity,
        agent_filter=agent_id,
    )
    total = health_monitor.get_events_total(
        severity_filter=severity,
        agent_filter=agent_id,
    )
    return {
        "events": [e.to_dict() for e in events],
        "total": total,
        "unacknowledged_count": health_monitor.get_unacknowledged_count(),
        "counts_by_severity": health_monitor.get_counts_by_severity(),
    }


@router.get("/status")
async def get_health_status() -> dict[str, Any]:
    """Get health status for all tracked agents."""
    details = health_monitor.get_all_health_details()

    summary = {"healthy": 0, "warning": 0, "critical": 0, "offline": 0}
    for agent_detail in details.values():
        status = agent_detail["status"]
        if status in summary:
            summary[status] += 1

    return {
        "agents": details,
        "summary": summary,
        "auto_recovery_enabled": health_monitor.config.auto_recovery_enabled,
        "auto_redeploy_on_reconnect": health_monitor.config.auto_redeploy_on_reconnect,
        "monitoring_active": health_monitor._running,
    }


@router.get("/status/{agent_id}")
async def get_agent_health_status(agent_id: UUID) -> dict[str, Any]:
    """Get health status for a specific agent."""
    detail = health_monitor.get_agent_health_detail(agent_id)
    if detail is None:
        raise NotFoundError(
            "Agent health data not found",
            details={"agent_id": str(agent_id)},
        )
    return {"agent_id": str(agent_id), **detail}


@router.post("/events/{event_id}/acknowledge")
async def acknowledge_event(event_id: str) -> dict[str, Any]:
    """Mark a health event as acknowledged."""
    found = health_monitor.acknowledge_event(event_id)
    if not found:
        raise NotFoundError(
            "Health event not found",
            details={"event_id": event_id},
        )
    return {"acknowledged": True, "event_id": event_id}


@router.delete("/events")
async def clear_events() -> dict[str, Any]:
    """Clear all health events."""
    count = health_monitor.clear_events()
    return {"cleared": count}


@router.get("/config")
async def get_config() -> dict[str, Any]:
    """Get current health monitor configuration."""
    return health_monitor.get_config_dict()


@router.put("/config")
async def update_config(updates: dict[str, Any]) -> dict[str, Any]:
    """Update health monitor configuration.

    Accepts a dict of config keys to update. Invalid keys are ignored.
    """
    return health_monitor.update_config(updates)
